MdeModulePkg/Universal/Console/TerminalDxe/TerminalConOut.c

```diff
@@ -562,96 +561,17 @@ TerminalConOutSetAttribute (
   //  convert Attribute value to terminal emulator
   //  understandable foreground color
   //
-  switch (Attribute & 0x07) {
-    case EFI_BLACK:
-      ForegroundControl = 30;
-      break;
-
-    case EFI_BLUE:
-      ForegroundControl = 34;
-      break;
-
-    case EFI_GREEN:
-      ForegroundControl = 32;
-      break;
-
-    case EFI_CYAN:
-      ForegroundControl = 36;
-      break;
-
-    case EFI_RED:
-      ForegroundControl = 31;
-      break;
-
-    case EFI_MAGENTA:
-      ForegroundControl = 35;
-      break;
-
-    case EFI_BROWN:
-      ForegroundControl = 33;
-      break;
-
-    default:
-
-    case EFI_LIGHTGRAY:
-      ForegroundControl = 37;
-      break;
-  }
-
-  //
-  //  bit4 of the Attribute indicates bright control
-  //  of terminal emulator.
-  //
-  BrightControl = (UINT8)((Attribute >> 3) & 1);
-
-  //
-  //  convert Attribute value to terminal emulator
-  //  understandable background color.
-  //
-  switch ((Attribute >> 4) & 0x07) {
-    case EFI_BLACK:
-      BackgroundControl = 40;
-      break;
-
-    case EFI_BLUE:
-      BackgroundControl = 44;
-      break;
-
-    case EFI_GREEN:
-      BackgroundControl = 42;
-      break;
-
-    case EFI_CYAN:
-      BackgroundControl = 46;
-      break;
-
-    case EFI_RED:
-      BackgroundControl = 41;
-      break;
-
-    case EFI_MAGENTA:
-      BackgroundControl = 45;
-      break;
-
-    case EFI_BROWN:
-      BackgroundControl = 43;
-      break;
-
-    default:
-
-    case EFI_LIGHTGRAY:
-      BackgroundControl = 47;
-      break;
-  }
+  ForegroundControl = mTerminalEfiColors[Attribute & 0x07];
+  BackgroundControl = mTerminalEfiColors[(Attribute >> 4) & 0x07];
 
   //
   // terminal emulator's control sequence to set attributes
   //
-  mSetAttributeString[BRIGHT_CONTROL_OFFSET]         = (CHAR16)('0' + BrightControl);
-  mSetAttributeString[FOREGROUND_CONTROL_OFFSET + 0] = (CHAR16)('0' + (ForegroundControl / 10));
-  mSetAttributeString[FOREGROUND_CONTROL_OFFSET + 1] = (CHAR16)('0' + (ForegroundControl % 10));
-  mSetAttributeString[BACKGROUND_CONTROL_OFFSET + 0] = (CHAR16)('0' + (BackgroundControl / 10));
-  mSetAttributeString[BACKGROUND_CONTROL_OFFSET + 1] = (CHAR16)('0' + (BackgroundControl % 10));
+  CHAR16 mSetTerminalColorString[64];
+  UnicodeSPrint(mSetTerminalColorString, sizeof(mSetTerminalColorString), 
+    L"%c[38;2;%u;%u;%um%c[48;2;%u;%u;%um",
+    ESC, ForegroundControl.r, ForegroundControl.g, ForegroundControl.b,
+    ESC, BackgroundControl.r, BackgroundControl.g, BackgroundControl.b);

```

Comment:
```
  mSetAttributeString[BRIGHT_CONTROL_OFFSET]         = (CHAR16)('0' + BrightControl);
  mSetAttributeString[FOREGROUND_CONTROL_OFFSET + 0] = (CHAR16)('0' + (ForegroundControl / 10));
  mSetAttributeString[FOREGROUND_CONTROL_OFFSET + 1] = (CHAR16)('0' + (ForegroundControl % 10));
  mSetAttributeString[BACKGROUND_CONTROL_OFFSET + 0] = (CHAR16)('0' + (BackgroundControl / 10));
  mSetAttributeString[BACKGROUND_CONTROL_OFFSET + 1] = (CHAR16)('0' + (BackgroundControl % 10));
```

This approach would work too, but you would have to change the mSetAttributeString string and adjust BACKGROUND_CONTROL_OFFSET, BRIGHT_CONTROL_OFFSET.